File: Recipe - Java/app/src/main/java/com/example/recipe_java/ui/foodstuff/FoodstuffFragment.java
package com.example.recipe_java.ui.foodstuff;

import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.ViewPager;
import com.example.recipe_java.Data;
import com.example.recipe_java.R;
import com.example.recipe_java.common.ViewPagerAdapter;
import com.example.recipe_java.models.Foodstuff;
import com.google.android.material.tabs.TabLayout;
import org.jetbrains.annotations.NotNull;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class FoodstuffFragment extends Fragment implements TabLayout.OnTabSelectedListener {

    private TabLayout tabLayout;
    private ViewPager viewPager;
    private TextView nameTextView;
    private EditText caloriesValue;
    private EditText proteinValue;
    private EditText fatsValue;
    private EditText sugarValue;
    private EditText vitaminAValue;
    private EditText vitaminCValue;
    private Button submitButton;
    private CheckBox checkBoxDiabetics;
    private RecyclerView foodstuffRecyclerView;
    private FoodstuffAdapter foodstuffAdapter;

    @Override
    public void onCreate(@Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @org.jetbrains.annotations.Nullable
    @Override
    public View onCreateView(@NonNull @NotNull LayoutInflater inflater, @Nullable @org.jetbrains.annotations.Nullable ViewGroup container, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_foodstuff, container, false);
    }

    @Override
    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        tabLayout = view.findViewById(R.id.tabLayout);
        viewPager = view.findViewById(R.id.viewPager);
        submitButton = view.findViewById(R.id.submitButton);
        nameTextView = view.findViewById(R.id.foodstuffName);
        caloriesValue = view.findViewById(R.id.caloriesValue);
        proteinValue = view.findViewById(R.id.proteinValue);
        fatsValue = view.findViewById(R.id.fatsValue);
        sugarValue = view.findViewById(R.id.sugarValue);
        vitaminAValue = view.findViewById(R.id.vitaminAValue);
        vitaminCValue = view.findViewById(R.id.vitaminCValue);
        checkBoxDiabetics = view.findViewById(R.id.isGoodForDiabeticsCheckBox);
        foodstuffRecyclerView = view.findViewById(R.id.foodstuffRecyclerView);

        initData();
        initListeners();
    }

    public void initData() {
        int[] layouts = new int[]{R.id.newFoodstuffConstraint, R.id.foodstuffListConstraint};
        viewPager.setAdapter(new ViewPagerAdapter(getContext(), layouts));

        viewPager.setOnPageChangeListener(new ViewPager.OnPageChangeListener(){

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                tabLayout.getTabAt(position).select();
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });

        List<String> tabTitles = Arrays.asList(getContext().getResources().getStringArray(R.array.foodstuff_tabs));
        for (String title: tabTitles) {
            tabLayout.addTab(
                    tabLayout.newTab()
                            .setText(title)
                            .setTag(title.toLowerCase())
            );
        }

        tabLayout.addOnTabSelectedListener(this);

        foodstuffRecyclerView.setAdapter(new FoodstuffAdapter());
        foodstuffAdapter = (FoodstuffAdapter) foodstuffRecyclerView.getAdapter();
        foodstuffAdapter.setFoodstuff(Data.foodstuff);
    }

    private void initListeners() {
        checkBoxDiabetics.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @RequiresApi(api = Build.VERSION_CODES.N)
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                if (checkBoxDiabetics.isChecked()) {
                    // Example: function collection - filter
                    List<Foodstuff> goodForDiabetics = Data.foodstuff.stream().filter(
                            Foodstuff::isGoodForDiabetics
                    ).collect(Collectors.toList());
                    foodstuffAdapter.setFoodstuff((ArrayList<Foodstuff>) goodForDiabetics);
                } else {
                    foodstuffAdapter.setFoodstuff(Data.foodstuff);
                }
            }
        });

        submitButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String name = nameTextView.getText().toString();
                Integer calories = Integer.parseInt(caloriesValue.getText().toString());
                Integer protein = Integer.parseInt(proteinValue.getText().toString());
                Integer fats = Integer.parseInt(fatsValue.getText().toString());
                Integer sugar = Integer.parseInt(sugarValue.getText().toString());
                Integer vitaminA = Integer.parseInt(vitaminAValue.getText().toString());
                Integer vitaminC = Integer.parseInt(vitaminCValue.getText().toString());
                Foodstuff foodstuff = new Foodstuff(name, calories, protein, fats, sugar, vitaminA, vitaminC);
                Data.foodstuff.add(foodstuff);
                Toast.makeText(getContext(), R.string.toast_foodstuff_added, Toast.LENGTH_LONG).show();
            }
        });
    }

    @Override
    public void onTabSelected(TabLayout.Tab tab) {
        viewPager.setCurrentItem(tab.getPosition());
    }

    @Override
    public void onTabUnselected(TabLayout.Tab tab) { }

    @Override
    public void onTabReselected(TabLayout.Tab tab) { }
}

File: Recipe - Kotlin/app/src/main/java/com/example/recipe/ui/foodstuff/FoodstuffFragment.kt
package com.example.recipe.ui.foodstuff

import ViewPagerAdapter
import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.ViewPager
import com.example.recipe.Data
import com.example.recipe.R
import com.google.android.material.tabs.TabLayout
import com.example.recipe.models.Foodstuff

class FoodstuffFragment : Fragment(), TabLayout.OnTabSelectedListener {

    private lateinit var tabLayout: TabLayout
    private lateinit var viewPager: ViewPager
    private lateinit var nameTextView: TextView
    private lateinit var caloriesValue: EditText
    private lateinit var proteinValue: EditText
    private lateinit var fatsValue: EditText
    private lateinit var sugarValue: EditText
    private lateinit var vitaminAValue: EditText
    private lateinit var vitaminCValue: EditText
    private lateinit var submitButton: Button
    private lateinit var checkBoxDiabetics: CheckBox
    private lateinit var foodstuffRecyclerView: RecyclerView

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val root = inflater.inflate(R.layout.fragment_foodstuff, container, false)

        tabLayout = root.findViewById(R.id.tabLayout)
        viewPager = root.findViewById(R.id.viewPager)
        submitButton = root.findViewById(R.id.submitButton)
        nameTextView = root.findViewById(R.id.foodstuffName)
        caloriesValue = root.findViewById(R.id.caloriesValue)
        proteinValue = root.findViewById(R.id.proteinValue)
        fatsValue = root.findViewById(R.id.fatsValue)
        sugarValue = root.findViewById(R.id.sugarValue)
        vitaminAValue = root.findViewById(R.id.vitaminAValue)
        vitaminCValue = root.findViewById(R.id.vitaminCValue)
        checkBoxDiabetics = root.findViewById(R.id.isGoodForDiabeticsCheckBox)
        foodstuffRecyclerView = root.findViewById(R.id.foodstuffRecyclerView)

        initData()
        initListeners()

        return root
    }

    @SuppressLint("SetTextI18n")
    private fun initData() {
        viewPager.adapter = ViewPagerAdapter(listOf(R.id.newFoodstuffConstraint, R.id.foodstuffListConstraint))
        viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrollStateChanged(state: Int) {}
            override fun onPageScrolled(
                position: Int,
                positionOffset: Float,
                positionOffsetPixels: Int
            ) {}

            override fun onPageSelected(position: Int) {
                tabLayout.getTabAt(position)?.select()
            }
        })

        val tabTitles = context!!.resources.getStringArray(R.array.foodstuff_tabs)
        for (title in tabTitles) {
            tabLayout.addTab(
                tabLayout.newTab()
                    .setText(title)
                    .setTag(title.lowercase())
            )
        }

        tabLayout.addOnTabSelectedListener(this)

        // Example: var and val
//        val buttonText1 = "Step"
//        var buttonText2 = 1
//        val buttonText1: String = "Step"
//        var buttonText2: Int = 1
//
//        buttonText1 = "Submit"
//        buttonText2 = 2
//
//        buttonText1++
//        buttonText2++
//
//        submitButton.text = buttonText1 + buttonText2

        foodstuffRecyclerView.adapter = FoodstuffAdapter()
        (foodstuffRecyclerView.adapter as FoodstuffAdapter).setFoodstuff(Data.foodstuff)

        checkBoxDiabetics.setOnCheckedChangeListener(CompoundButton.OnCheckedChangeListener { buttonView, isChecked ->
            if (isChecked) {
                // Example: function collection - filter
                val goodForDiabetics = Data.foodstuff.filter {
                    return@filter it.isGoodForDiabetics()
                }.toMutableList()
                (foodstuffRecyclerView.adapter as FoodstuffAdapter).setFoodstuff(goodForDiabetics)
            } else {
                (foodstuffRecyclerView.adapter as FoodstuffAdapter).setFoodstuff(Data.foodstuff)
            }
        })
    }

    private fun initListeners() {
        submitButton.setOnClickListener {
            val name = nameTextView.text.toString()
            val calories = caloriesValue.text.toString().toInt()
            val protein = proteinValue.text.toString().toInt()
            val fats = fatsValue.text.toString().toInt()
            val sugar = sugarValue.text.toString().toInt()
            val vitaminA = vitaminAValue.text.toString().toInt()
            val vitaminC = vitaminCValue.text.toString().toInt()
            val foodstuff = Foodstuff(name, calories, protein, fats, sugar, vitaminA, vitaminC)
            Data.foodstuff.add(foodstuff)
            Toast.makeText(context, R.string.toast_foodstuff_added, Toast.LENGTH_LONG).show()
        }
    }

    override fun onTabSelected(tab: TabLayout.Tab?) {
        viewPager.currentItem = tab?.position ?: 0
    }

    override fun onTabUnselected(tab: TabLayout.Tab?) {
        Log.d("TAB UNSELECTED", tab?.tag as String)
    }

    override fun onTabReselected(tab: TabLayout.Tab?) {
        Log.d("TAB RESELECTED", tab?.tag as String)
    }
}
